File: bindings/python/examples/srcml_copy_archive.py
#!/usr/bin/python
#
# @file srcml_copy_archive.py
#
# @copyright Copyright (C) 2013-2014 srcML, LLC. (www.srcML.org)
#
# The srcML Toolkit is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# The srcML Toolkit is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the srcML Toolkit; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 


# Example program that uses the python API.
# Copy an archive.


import sys, srcml

if __name__ == "__main__":
    # int i;
    iarchive = None
    oarchive = None
    unit = None

     # open up an existing archive 
    iarchive = srcml.create_archive()
    srcml.read_open_filename(iarchive, "main.cpp.xml")

    # create a new srcml archive structure 
    # options and attributes of cloned archive start the same as
    #   the original archive 
    oarchive = srcml.clone_archive(iarchive)

    # open a srcML archive for output 
    srcml.write_open_filename(oarchive, "cloned_archive.xml")

    # copy the files from the input archive to the output archive 
    unit = srcml.read_unit(iarchive)
    while unit != None:
        # Translate to srcml and append to the archive 
        srcml.write_unit(oarchive, unit)

        srcml.free_unit(unit)
        unit = srcml.read_unit(iarchive)

    # close the archives 
    srcml.close_archive(oarchive)
    srcml.close_archive(iarchive)

    # free the srcML archive data 
    srcml.free_archive(oarchive)
    srcml.free_archive(iarchive)

File: examples/libsrcml/srcml_copy_archive.cpp
// SPDX-License-Identifier: GPL-3.0-only
/**
 * @file srcml_copy_archive.cpp
 *
 * @copyright Copyright (C) 2013-2019 srcML, LLC. (www.srcML.org)
 *
 * Example program of the use of the libsrcml C API.
 *
 * Copy an archive.
*/

#include <srcml.h>

int main(int argc, char* argv[]) {

    // open up an existing archive
    srcml_archive* inputArchive = srcml_archive_create();
    srcml_archive_read_open_filename(inputArchive, "project.xml");

    // create a new srcml archive structure
    // options and attributes of cloned archive start the same
    // original archive
    srcml_archive* outputArchive = srcml_archive_clone(inputArchive);

    // open a srcML archive for output
    srcml_archive_write_open_filename(outputArchive, "projectCopy.xml");

    // copy the files from the input archive to the output archive
    srcml_unit* unit = nullptr;
    while ((unit = srcml_archive_read_unit(inputArchive))) {

        // append to the output archive
        srcml_archive_write_unit(outputArchive, unit);

        srcml_unit_free(unit);
    }

    // close the archives
    srcml_archive_close(outputArchive);
    srcml_archive_close(inputArchive);

    // free the archives
    srcml_archive_free(outputArchive);
    srcml_archive_free(inputArchive);

    return 0;
}
